src/LiquidityPool.sol

```diff
@@ -93,13 +103,44 @@ contract LiquidityPool is Auth, IERC7540 {
 
     // --- ERC-7540 methods ---
     /// @inheritdoc IERC7540Deposit
-    function requestDeposit(uint256 assets, address receiver) public {
-        require(IERC20(asset).balanceOf(msg.sender) >= assets, "LiquidityPool/insufficient-balance");
+    function requestDeposit(uint256 assets, address receiver, address owner, bytes memory data)
+        public
+        returns (uint256)
+    {
+        require(owner == msg.sender, "LiquidityPool/not-msg-sender");
+        require(IERC20(asset).balanceOf(owner) >= assets, "LiquidityPool/insufficient-balance");
+
+        require(manager.requestDeposit(address(this), assets, receiver, owner), "LiquidityPool/request-deposit-failed");
+        SafeTransferLib.safeTransferFrom(asset, owner, address(escrow), assets);
+
         require(
-            manager.requestDeposit(address(this), assets, receiver, msg.sender), "LiquidityPool/request-deposit-failed"
+            data.length == 0 || receiver.code.length == 0
```

Comment:
The callback is only called if nonzero data was passed, and the receiver is a contract.

There should not be any reentrancy issues because the callback occurs after all state changes (only the event emission and the return call happen after this)